src/editor/codemirror/copypaste.ts

```diff
@@ -0,0 +1,78 @@
+import { Extension } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import { lineNumFromUint8Array } from "../../common/text-util";
+import { deployment } from "../../deployment";
+import { CodeInsertType } from "./dnd";
+import { calculateChanges } from "./edits";
+
+interface PasteContext {
+  code: string;
+  type: CodeInsertType;
+}
+
+const getCodeFromHtml = (
+  innerHTML: string | undefined
+): PasteContext | undefined => {
+  if (!innerHTML) {
+    return;
+  }
+  const parent: HTMLElement = document.createElement("html");
+  parent.innerHTML = innerHTML;
+  return {
+    code: (parent as HTMLElement).querySelector("code")?.textContent || "",
+    type:
+      ((parent as HTMLElement)
+        .querySelector("code")
+        ?.getAttribute("data-type") as CodeInsertType) || "unknown",
+  };
+};
+
+const copyPasteHandlers = () => {
+  const textEncoder = new TextEncoder();
+  return [
+    EditorView.domEventHandlers({
+      paste(event, view) {
+        if (
+          !view.state.facet(EditorView.editable) ||
+          event.clipboardData?.getData("text")
+        ) {
+          return;
+        }
+        event.preventDefault();
+        const pasteContext = getCodeFromHtml(
+          event.clipboardData?.getData("text/html")
+        );
+        if (!pasteContext) {
+          return;
+        }
+        // Should we use lineCount here, or follow the dnd logging?
+        // If we use lineCount, should we ignore imports and empty lines?
+        const lineCount = lineNumFromUint8Array(
+          // Ignore leading/trailing lines.
+          textEncoder.encode(pasteContext.code.trim())
+        );
+        deployment.logging.event({
+          type: "paste",
+          value: lineCount,
+        });
+
+        const lineNumber = view.state.doc.lineAt(
+          view.state.selection.ranges[0].from
+        ).number;
+
+        view.dispatch(
```

Comment:
Is it possible to tag this transaction as "input.paste" so that the normal line count code runs? 

Or perhaps we shouldn't even run it, as it's not desirable to have an event that conflates regular pastes and reference/api pastes. Let's discuss this.

Reply:
It might be better to log it in the same we do for dnd drop events.

Reply:
Discussed. Let's log it analogously to dnd and not log it as a user paste at all.

Reply:
[`738ca6a`](https://github.com/microbit-foundation/python-editor-next/pull/766/commits/738ca6a4205e75ce2349f6b8ace0eeb6c333c4a2). A user paste is logged as `paste` with number of lines, and, as with all of the logging events that handle code examples, the special code paste is logged as `code-paste` with the code id.